### DISCUSSION TITLE
I built a brainfuck CPU

### DISCUSSION THREAD

**Discussion by gbizzotto** [hneemann/Digital#588] [Show and tell]
I've spent the last few weeks building [this](https://github.com/gbizzotto/digital).

I'm just happy it works.

You might need to clone the whole repo (which is small) and set Digital's `Library` path for it to be able to find all the parts.

Take a look at the circuits cited in the repo's description.

Here's the "module" version, with external memory to store the program.

![Screen Shot 2020-12-21 at 21 20 44](https://user-images.githubusercontent.com/5064376/102834321-7c262580-43d2-11eb-908f-4fc6bd4da11b.png)

Here's the standalone version, with its own program memory.

![Screen Shot 2020-12-21 at 21 28 03](https://user-images.githubusercontent.com/5064376/102834688-8c8ad000-43d3-11eb-9261-9600cc2ee9b0.png)

**Comment by hneemann** [comment]
Your display is really impressive! :+1: 
Even the registers are built from simple gates!

**Reply by gbizzotto** [reply]
Copy/paste magic!

**Reply by hneemann** [reply]
But the courage to try to simulate hundreds of thousands of gates with Digital! :smile:

**Reply by gbizzotto** [reply]
What would it take to have a native backend for running simulations?
I could help with C/C++.
I managed to compile and run Digital from source. I'll take a look at that part, see if there's some low hanging optimization fruit.

**Reply by hneemann** [reply]
I don't think that there is much to get with C++ here. The JIT compiler works pretty well these days.
And the algorithm used for the simulation is already quite optimized. 
One could do without the simulation of gate delays, but then some circuits would no longer work. For example, the differentiation of signals would be affected: A&!A would always be zero. In Digital, however, a short high pulse is generated at the rising edge of A, because the NOT has a delay. And this behavior is required to illustrate the function of edge-triggered flip-flops.

**Reply by gbizzotto** [reply]
I get 2 orders of magnitude of slowdown when i use my ridiculous RGB display instead of the built-in terminal.
It seems odd to me, since signals only go through it occasionally. Most of the time nothing happens in it.
I understand it uses up loads of RAM, but i thought Digital would only care about (and spend time) propagating phase changes where it matters, and not be affected by the other stuff that's basically turned off.

**Reply by hneemann** [reply]
When I run the simulation with the RGB LED display, the simulation runs at 1kHz and after a view seconds the speed goes up to 100kHz. 
But there is nothing to see on the display.

**Reply by gbizzotto** [reply]
I'm not sure whether the 1kHz phase is the preprocessor compacting the code, or java JITting everything.
Each pixel takes about a minute to compute (compare to a few seconds with a terminal)
You should be able to see the the wire pointing at row changing for every pixel.
Yeah, it's slow af. I'll work on that.

**Reply by hneemann** [reply]
I played around a bit with a display that is made of single LEDs. 
I found and fixed some performance problems with the model creation at the simulation start. 
Please use the latest [pre-release](https://infdigital.dhbw-mosbach.de/) or the initSpeedUp branch.

The example uses 16384 RGB-LEDs and a 6 bit D-flip-flop per pixel as a buffer:

https://user-images.githubusercontent.com/7190563/103148171-49c35200-475d-11eb-89f2-8cf47ff51369.mp4

**Reply by gbizzotto** [reply]
Fantastic! You squeezed out more than an order of magnitude there!

**Reply by hneemann** [reply]
A very silly mistake: At two critical spots I used a list search instead of a hash set lookup. 
Not noticeable when the number of components in a circuit is small, but a huge difference when the number becomes large.

**Reply by gbizzotto** [reply]
I'm redrawing that RGB LEDs wall with that pre-release. All I did was resize the LED from 1 to 2, so it covers the underlying module. It's probably what you did in the video above. Anyway, i'm trying to paste a bloc of 4 rows (of 132) and it's taking forever. The first block took seconds. The fourth took maybe 10 minutes. Not sure if I'll be able to complete it.

The strange this is, the stable version I used to create the first version of the RGB wall was not particularly responsive due to the sheer size of the thing, but it was manageable.
Would the size of the LED somehow impact performance? Or any other modification?

This screenshot shows (so you have an idea of the size) the pasted block not getting deselected after the location has been chosen. The app is now unresponsive. I guess it'll snap back eventually.
![Screen Shot 2020-12-27 at 18 35 39](https://user-images.githubusercontent.com/5064376/103180210-656c4c80-4872-11eb-9c2c-dd30cd076bb2.png)

**Reply by hneemann** [reply]
There is indeed a problem when a large number of diagonal wires are inserted.
I suspect that there are many diagonal wires in this very compact circuit?

**Reply by gbizzotto** [reply]
I'd say about 15000, so I guess yes.
The release version works fine. I only had that problem with the pre-release.

**Reply by hneemann** [reply]
Yes, I noticed that in your example circuit there are many diagonal wires that are present multiple times, which makes no sense.
I have added code that removes these duplicate diagonal wires, which causes this new problem.

**Reply by gbizzotto** [reply]
By "present multiple times" you mean two wires on top of each other, with identical starting and ending points?
If so, i thought Digital would remove them when they get pasted.
If you mean the repeating pattern present in the design, it's how I managed to have a compact 4x4 LED pattern that I could copy and paste around. I couldn't have straight wires crossing a LED pin without contact.
I could do it all with tunnels. Would that be easier on the software?

**Reply by hneemann** [reply]
> By "present multiple times" you mean two wires on top of each other, with identical starting and ending points?
> If so, i thought Digital would remove them when they get pasted.

Yes, but it didn't work with diagonal wires in v0.25. And removing them makes the insertion so slow.
But I was able to fix that. Please try the latest [pre-release](https://infdigital.dhbw-mosbach.de/).

**Reply by gbizzotto** [reply]
Ah yes, copying and pasting large chunks is MUCH faster. Thanks (:

**Comment by gbizzotto** [comment]
I've wired a preprocessor that transforms the brainfuck code so that the actual execution becomes easy as pie. I got 2 orders of magnitude out of it and could run `mandelbrot.b` with an RGB display in about 3 hours.

![Screen Shot 2020-12-28 at 07 59 36](https://user-images.githubusercontent.com/5064376/103209957-190d2500-48e3-11eb-8161-f8cc63844278.png)

https://user-images.githubusercontent.com/5064376/103210077-65586500-48e3-11eb-85bb-13968b35c8b2.mov
